dpki/src/keypair.rs

```diff
@@ -174,6 +174,33 @@ impl EncryptingKeyPair {
     }
 
     // TODO: Encrypt and decrypt functions
+
+    /// encrypt some arbitrary data with the signing private key
+    /// @param {SecBuf} data - the data to sign
+    /// @return {SecBuf} signature - Empty SecBuf to be filled with the signature
+    pub fn encrypt(&mut self, data: &mut SecBuf) -> HcResult<SecBuf> {
+        let mut nonce = SecBuf::with_insecure(lib3h_sodium::aead::NONCEBYTES);
+        let mut cipher = SecBuf::with_insecure(data.len() + lib3h_sodium::aead::ABYTES);
+        lib3h_sodium::aead::enc(data, &mut self.private, None, &mut nonce, &mut cipher)?;
+        Ok(cipher.clone())
+    }
+
+    /// encrypt some arbitrary data with the signing private key
+    /// @param {SecBuf} data - the data to sign
+    /// @return {SecBuf} signature - Empty SecBuf to be filled with the signature
+    pub fn decrypt(&mut self, cipher: &mut SecBuf) -> HcResult<SecBuf> {
```

Comment:
I think we need the output to be a parameter, so that the implementer can decide if the data need to be protected or not.

something like:

```rust
pub fn decrypt(&mut self, cipher: &mut SecBuf, result: &mut SecBuf) -> HcResult<()> {
```

``` rust
let res = SecBuf::with_secure(cipher.len() + ABYTES);
store.decrypt(&mut cipher, &mut res)?;
```

For symmetry, we could also make the output of encrypt be a param, but that's not strictly needed, since the encrypted data shouldn't need to be protected.

Reply:
The reason I implemented it this way was because we have a few unsafe spots deep in the encryption code and thought it would be better to reduce the area of mutability by just returning the new encrypted data. That way that way they have a new cleaner block of data that minimizes the side effects

Reply:
I don't understand this reasoning. The unsafe code is executed in both cases, either by us or by the user of this api. We need to make sure that is solid through whatever testing is needed, but is also not related to my comment above.

The current code does not allow the user to memory protect the results of the decryption. If they are decrypting sensitive data from somewhere, they may need to use `with_secure` instead of `with_insecure` which the current api does not allow.

Reply:
I guess the comment was related to the output type itself, I would still like to output a new piece of data just to be a bit closer to determinism but i guess it is okay since rust compiler will always take care of shared mutability. 

How do you feel about an enum just to make it more rust style?

decrypt(&mut self, cipher,CryptoOptions::Secure) -> HcResult<String>

and crypto options could be Secure and InSecure?

Reply:
I swear I'm not trying to be argumentative, lol : ) But, I don't see how the determinism is different in either case, and I don't like the dissonance of implementing the HDK api in a different manner than the underlying crypto api.

I feel like I'm missing some context for why this is a problem. Is this specifically because of the unsafe copy in the write function: `std::ptr::copy(data.as_ptr(), (**b).as_mut_ptr().add(offset), data.len());`?

I misspoke earlier when I said that would happen in either case. You are correct, it doesn't get called when creating a new SecBuf. That code was originally a for-loop copy of the bytes, but that can't take advantage of memory block size optimizations in the processor, the ptr copy is far more efficient, and there is an explicit bonds check above it, so there shouldn't be any real unsafety (though if you think of additional unit tests needed, please feel free to add them : )

Ultimately, though, even rust std code uses mem copy techniques at its lowest levels, if there's an efficiency reason to do it, I don't feel like we should fear diving in.

I'm just making things up at this point... please let me know if I'm totally off the mark. Thanks!

Reply:
I have made the changes though and even though I have my opinions on rust and unsafety, you have worked on this side of the code more and I trust your expertise!